src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs

```diff
@@ -0,0 +1,932 @@
+                }
+            }
+            else
+#endif
+            {
+                for (int i = 0; i < secret.Length; i += 16)
+                {
+                    BinaryPrimitives.WriteUInt64LittleEndian(
+                        destinationSecret.Slice(i),
+                        BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(i)) + seed);
+
+                    BinaryPrimitives.WriteUInt64LittleEndian(
+                        destinationSecret.Slice(i + 8),
+                        BinaryPrimitives.ReadUInt64LittleEndian(secret.Slice(i + 8)) - seed);
+                }
+            }
+        }
+
+        /// <summary>Loops over <see cref="Accumulate512"/>.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Accumulate(Span<ulong> acc, ReadOnlySpan<byte> source, ReadOnlySpan<byte> secret, int stripesPerBlock)
+        {
+            for (int i = 0; i < stripesPerBlock; i++)
+            {
+                Accumulate512(acc, source.Slice(i * StripeLengthBytes), secret.Slice(i * SecretConsumeRateBytes));
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Accumulate512(Span<ulong> acc, ReadOnlySpan<byte> source, ReadOnlySpan<byte> secret)
+        {
+            Debug.Assert(acc.Length == 8);
+            Debug.Assert(secret.Length >= StripeLengthBytes);
+
+#if NET7_0_OR_GREATER
+            if (Vector256.IsHardwareAccelerated && BitConverter.IsLittleEndian)
+            {
+                ref ulong accRef = ref MemoryMarshal.GetReference(acc);
+                ref ulong accEnd = ref Unsafe.Add(ref accRef, acc.Length);
+
+                ref uint sourceRef = ref Unsafe.As<byte, uint>(ref MemoryMarshal.GetReference(source));
+                ref uint keyRef = ref Unsafe.As<byte, uint>(ref MemoryMarshal.GetReference(secret));
+
+                while (Unsafe.IsAddressLessThan(ref accRef, ref accEnd))
+                {
+                    Vector256<ulong> accVec = Vector256.LoadUnsafe(ref accRef);
+                    Vector256<uint> sourceVec = Vector256.LoadUnsafe(ref sourceRef);
+                    Vector256<uint> sourceKey = sourceVec ^ Vector256.LoadUnsafe(ref keyRef);
+
+                    // TODO: Figure out how to unwind this shuffle and just use Vector256.Multiply
+                    Vector256<uint> sourceKeyLow = Vector256.Shuffle(sourceKey, Vector256.Create(1u, 0, 3, 0, 5, 0, 7, 0));
+                    Vector256<uint> sourceSwap = Vector256.Shuffle(sourceVec, Vector256.Create(2u, 3, 0, 1, 6, 7, 4, 5));
+                    Vector256<ulong> sum = accVec + sourceSwap.AsUInt64();
+                    Vector256<ulong> product = Avx2.IsSupported ?
+                        Avx2.Multiply(sourceKey, sourceKeyLow) :
+                        (sourceKey & Vector256.Create(~0u, 0u, ~0u, 0u, ~0u, 0u, ~0u, 0u)).AsUInt64() * (sourceKeyLow & Vector256.Create(~0u, 0u, ~0u, 0u, ~0u, 0u, ~0u, 0u)).AsUInt64();
```

Comment:
@tannergooding is `operator *` for `Vector128<ulong>` supported on ARM? judging by the codegen it is not (fallbacks to software impl)